File: system/rust/src/types/address.rs
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::fmt;

/// Represents Bluetooth address.
#[repr(transparent)]
#[derive(Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Address {
    /// Internal representation.
    value: u64,
}

// SAFETY: The memory layout of the Rust `Address` struct matches the
// memory layout of the `ffi::Address` struct in C++ (a single uint64_t).
unsafe impl cxx::ExternType for Address {
    type Id = cxx::type_id!("ffi::Address");
    type Kind = cxx::kind::Trivial;
}

impl Address {
    /// Creates an Address from big-endian bytes (6 bytes).
    pub fn from_be_bytes(bytes: [u8; 6]) -> Self {
        let val = ((bytes[0] as u64) << 40)
            | ((bytes[1] as u64) << 32)
            | ((bytes[2] as u64) << 24)
            | ((bytes[3] as u64) << 16)
            | ((bytes[4] as u64) << 8)
            | (bytes[5] as u64);
        Self { value: val }
    }

    /// Returns the address as big-endian bytes (6 bytes).
    pub fn to_be_bytes(&self) -> [u8; 6] {
        [
            (self.value >> 40) as u8,
            (self.value >> 32) as u8,
            (self.value >> 24) as u8,
            (self.value >> 16) as u8,
            (self.value >> 8) as u8,
            self.value as u8,
        ]
    }

    /// Returns the full, non-anonymized address string (e.g., "AA:BB:CC:DD:EE:FF").
    /// Use this for JNI or internal system calls where the real address is required.
    pub fn to_full_string(&self) -> String {
        format!("{:?}", self)
    }
}

/// Anonymized display format: XX:XX:XX:XX:aa:bb
impl fmt::Display for Address {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let b = self.to_be_bytes();
        write!(f, "XX:XX:XX:XX:{:02X}:{:02X}", b[4], b[5])
    }
}

/// Detailed debug format: Shows full address.
/// Use of this for user-facing logs is discouraged.
impl fmt::Debug for Address {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let b = self.to_be_bytes();
        write!(f, "{:02X}:{:02X}:{:02X}:{:02X}:{:02X}:{:02X}", b[0], b[1], b[2], b[3], b[4], b[5])
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_address_display_anonymized() {
        let addr = Address::from_be_bytes([0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF]);
        assert_eq!(format!("{}", addr), "XX:XX:XX:XX:EE:FF");
    }

    #[test]
    fn test_address_debug_full() {
        let addr = Address::from_be_bytes([0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF]);
        assert_eq!(format!("{:?}", addr), "AA:BB:CC:DD:EE:FF");
    }

    #[test]
    fn test_endian_explicit_conversion() {
        let bytes = [0x01, 0x02, 0x03, 0x04, 0x05, 0x06];
        let addr = Address::from_be_bytes(bytes);
        assert_eq!(addr.to_be_bytes(), bytes);
    }
}

File: system/types/src/address.cc
/******************************************************************************
 *
 *  Copyright 2017 The Android Open Source Project
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at:
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ******************************************************************************/

#include <bluetooth/types/address.h>

#include <algorithm>
#include <array>
#include <cstdint>
#include <format>
#include <sstream>

static_assert(sizeof(RawAddress) == 6, "RawAddress must be 6 bytes long!");

const RawAddress RawAddress::kAny("ff:ff:ff:ff:ff:ff");
const RawAddress RawAddress::kEmpty("00:00:00:00:00:00");

std::string RawAddress::ToString() const {
  return std::format("{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}", address[0], address[1],
                     address[2], address[3], address[4], address[5]);
}

std::string RawAddress::ToRedactedStringForLogging() const {
  if (*this == RawAddress::kAny || *this == RawAddress::kEmpty) {
    return ToString();
  }

  return std::format("xx:xx:xx:xx:{:02x}:{:02x}", address[4], address[5]);
}

uint64_t RawAddress::ToUint64() const {
  return (static_cast<uint64_t>(address[0]) << 40) | (static_cast<uint64_t>(address[1]) << 32) |
         (static_cast<uint64_t>(address[2]) << 24) | (static_cast<uint64_t>(address[3]) << 16) |
         (static_cast<uint64_t>(address[4]) << 8) | (static_cast<uint64_t>(address[5]));
}

RawAddress RawAddress::FromUint64(uint64_t addr) {
  return RawAddress(std::array<uint8_t, 6>{
          static_cast<uint8_t>(addr >> 40),
          static_cast<uint8_t>(addr >> 32),
          static_cast<uint8_t>(addr >> 24),
          static_cast<uint8_t>(addr >> 16),
          static_cast<uint8_t>(addr >> 8),
          static_cast<uint8_t>(addr),
  });
}

std::optional<RawAddress> RawAddress::FromString(const std::string& from) {
  if (from.length() != 17) {
    return std::nullopt;
  }

  std::array<uint8_t, 6> address;
  std::istringstream stream(from);
  std::string token;
  int index = 0;

  while (getline(stream, token, ':')) {
    if (index >= 6) {
      return std::nullopt;
    }

    if (token.length() != 2) {
      return std::nullopt;
    }

    char* temp = nullptr;
    address[index] = std::strtol(token.c_str(), &temp, 16);

    if (temp == token.c_str()) {
      // string token is empty or has wrong format
      return std::nullopt;
    }

    if (temp != (token.c_str() + token.size())) {
      // cannot parse whole string
      return std::nullopt;
    }

    index++;
  }

  if (index != 6) {
    return std::nullopt;
  }

  return RawAddress(address);
}

RawAddress RawAddress::FromOctets(const uint8_t* from) {
  std::array<uint8_t, 6> address;
  std::copy(from, from + kLength, address.data());
  return RawAddress(address);
}

bool RawAddress::IsValidAddress(const std::string& address) {
  return RawAddress::FromString(address).has_value();
}
